### PR TITLE
model: Generate copyright statements from authors

### PR DESCRIPTION
**Pull request by bs-ondem**
Extend LicenseInfoResolver to generate copyright statements for declared
licenses based on author information.

### CODE REVIEW

File: model/src/main/kotlin/licenses/LicenseInfoResolver.kt [hunk 1]
```diff
@@ -94,6 +95,22 @@ class LicenseInfoResolver(
...
+                        matchingPathExcludes = emptyList(),
+                        copyrights = licenseInfo.declaredLicenseInfo.authors.mapTo(mutableSetOf()) {
+                            ResolvedCopyrightFinding(
+                                statement = "$COPYRIGHT_PREFIX $it",
+                                location = UNDEFINED_TEXT_LOCATION,
+                                matchingPathExcludes = emptyList()
+                            )
+                        }
+                    ))
+                }
```
Comment (fviernau) [inline model/src/main/kotlin/licenses/LicenseInfoResolver.kt:113]: This introduces "artifical" resolved license locations which in turn might introduce some "magic" into the code consuming the instance as soon as they want to handle the authors a bit different. What I could imagine e.g. is that one would want the NOTICE template to list the author copyrights separate from the licenses. This is why I believe that 

(1) the license info resolver should have a function to get the authors for a package.
(2) the `FreemarkerTemplateProcessor.PackageModel` should be extended by a property for the authors
(3) the notice template should be extended to iterate over the authors and insert them as needed, and thus
(4) The "Copyright (C)" should also go into the template, or at least into the reporter.

**Review by fviernau — CHANGES_REQUESTED** [review]

File: model/src/test/kotlin/licenses/DefaultLicenseInfoProviderTest.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+package org.ossreviewtoolkit.model.licenses
+
+import io.kotest.core.spec.style.WordSpec
+import io.kotest.matchers.shouldBe
+
+import org.ossreviewtoolkit.model.utils.SimplePackageConfigurationProvider
+
+class DefaultLicenseInfoProviderTest : WordSpec() {
+    init {
```
Comment (sschuberth) [inline model/src/test/kotlin/licenses/DefaultLicenseInfoProviderTest.kt:28]: Please use `WordSpec({...})` syntax to reduce indentation.

File: model/src/main/kotlin/licenses/LicenseInfoResolver.kt [hunk 3]
```diff
@@ -248,3 +265,6 @@ private class ResolvedLicenseBuilder(val license: SpdxSingleLicenseExpression) {
 
     fun build() = ResolvedLicense(license, originalDeclaredLicenses, originalExpressions, locations)
 }
+
+private val UNDEFINED_TEXT_LOCATION = TextLocation("/", TextLocation.UNKNOWN_LINE, TextLocation.UNKNOWN_LINE)
```
Comment (sschuberth) [inline model/src/main/kotlin/licenses/LicenseInfoResolver.kt:269]: Consider using "." instead of "/" which at least does not point to the file system root if something goes wrong 😉

File: model/src/main/kotlin/licenses/LicenseInfoResolver.kt (hunk 1 above)
Comment (fviernau) [reply model/src/main/kotlin/licenses/LicenseInfoResolver.kt:113]: @mnonnenmacher @sschuberth any opinions on the design considerations above?

Comment (sschuberth) [reply model/src/main/kotlin/licenses/LicenseInfoResolver.kt:113]: Aren't copyrights also visible in the web app report? Or isn't that at least planned? Anyway, independently of where copyrights are shown *right now*, I agree that this should be solved at the data model level, not at the reporter level, to be independent of the report format and any further processing. Such further processing could e.g. be the discussed `copyrightFindingCurations` that analogously to `licenseFindingCurations` could be added to `PackageConfiguration` at some point.

TL;DR while I agree that the `UNDEFINED_TEXT_LOCATION` is not nice, this is the best we can do for now without significant changes to the data model. I'm fine with the changes.

Comment (sschuberth) [reply model/src/main/kotlin/licenses/LicenseInfoResolver.kt:113]: > I proposed this to actually solve the problem via the data model

Well, by the name, yes. But it's the `FreemarkerTemplateProcessor`-specific data model, not the "real" / global ORT data model.

> without mixing copyright findings with authors.

At least in my view the goal always was to *do* turn authors into (some kind of) copyright findings; these copyrights were just not found by the scanner, but the analyzer.

It's just a bit unfortunate that we do not formally distinguish between declared and detected copyrights, like we do for licenses, and have an intermediate step that turn authors into declared copyrights.

**Comment by bs-ondem** [comment]
@sschuberth I think the WindowsTest is failing due to `PostgresStorageFunTest` exception what you have mentioned in the ORT Slack chat.

**Comment by sschuberth** [comment]
> @sschuberth I think the WindowsTest is failing due to `PostgresStorageFunTest` exception what you have mentioned in the ORT Slack chat.

I'm retriggering the check.

File: model/src/main/kotlin/licenses/LicenseInfoResolver.kt (hunk 1 above)
Comment (tsteenbe) [reply model/src/main/kotlin/licenses/LicenseInfoResolver.kt:113]: Sorry. only following the discussion half as I am not familiar with this bit of Kotlin code/Ort mofel but it looks to me that we artificially including authors in detected findings by giving it a fake text location and prefix it with "Copyright". I believe this not the right way to do this - prefer @fviernau to leave it up to reporter to choose how to display authors.

In the WebApp I was thinking to merger `Details` and `Licenses` sections into just `Detials` and the `authors` under the  a `Declared` sub-section

Comment (sschuberth) [reply model/src/main/kotlin/licenses/LicenseInfoResolver.kt:113]: As agreed in the developer meeting I've created an [issue](https://github.com/oss-review-toolkit/ort/issues/3780) to improve this later.

**Review by sschuberth — APPROVED** [review]
